lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -551,10 +556,129 @@ def build_module(self, trans, payload={}):
             'config_form'       : module.get_config_form(),
             'post_job_actions'  : module.get_post_job_actions(inputs)
         }
+        
+    @expose_api
+    def get_tool_predictions(self, trans, topk=10, max_seq_len=25, payload={}):
+        """
+        POST /api/workflows/get_tool_predictions
+        Fetch predicted tools for a workflow
+        """
+        trained_model_path = trans.app.config.model_path
+        tool_sequence = payload.get('tool_sequence', "")
+        if 'tool_sequence' not in payload or trained_model_path is None:
+            return
+        else:
+            r_file = requests.get(trained_model_path)
+            model_save_path = os.path.join(os.getcwd(), "database/trained_model.hdf5")
```

Comment:
Too many requests here at once could lead to multiple worker threads overwriting each other and result in corrupt hdf5 files? (Since you're opening + writing, rather and atomically moving.)

Reply:
The `hdf5` model file is read-only. There is a separate script to create it. For showing tool predictions, we just need to read the content of this model file.

Reply:
It wasn't read only though, it was written to disk in exactly this step.